### PR TITLE
Create a framework of functional tests for configuration files

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |
| ✓   | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

## Description

This create a framework for configuration testing.  Also migrate three standard unittest to the new framework for testing. Done to help with #4720.

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
It look like that enabling a message does not remove it from the disable list after the configuration parsing (see failing test). It must be handled later on. I think it will have to be fixed in pylint before merging.

File: pylint/testutils/configuration_test.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+
+EXPECTED_CONF_APPEND_KEY = "functional_append"
+EXPECTED_CONF_REMOVE_KEY = "functional_remove"
+EXPECTED_CONF_SPECIAL_KEYS = [EXPECTED_CONF_APPEND_KEY, EXPECTED_CONF_REMOVE_KEY]
+
+
+def get_expected_configuration(
+    configuration_path: str, default_configuration: Dict[str, Any]
+) -> Dict[str, Any]:
```
Comment (DanielNoord) [inline pylint/testutils/configuration_test.py:41]: Similar to previous comment: explanation for `Any` and `type: ignore`.

File: pylint/testutils/configuration_test.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    """Get the expected parsed configuration of a configuration functional test"""
+    result = copy.deepcopy(default_configuration)
+    config_as_json = get_expected_or_default(
+        configuration_path, suffix="result.json", default="{}"
+    )
+    to_override = json.loads(config_as_json)
+    for key, value in to_override.items():
+        if key not in EXPECTED_CONF_SPECIAL_KEYS:
+            result[key] = value
+    if EXPECTED_CONF_APPEND_KEY in to_override:
```
Comment (DanielNoord) [inline pylint/testutils/configuration_test.py:51]: Can't this be done in the `for` loop with an `elif`?
Similar for the `if` on L54.

File: pylint/testutils/configuration_test.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+def get_expected_output(configuration_path: str) -> Tuple[int, str]:
+    """Get the expected output of a functional test."""
+
+    def get_relative_path(path: str) -> str:
+        """Get the relative path we want without user specific information"""
+        return path.replace(USER_SPECIFIC, "")[1:]
+
+    output = get_expected_or_default(configuration_path, suffix="out", default="")
+    if output:
+        logging.info(
```
Comment (DanielNoord) [inline pylint/testutils/configuration_test.py:73]: Similar to other comment: do we normally use `logging` in functional tests?

File: pylint/testutils/configuration_test.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    )
+
+
+def run_using_a_configuration_file(
+    configuration_path: Union[Path, str], file_to_lint: str = __file__
+) -> Tuple[Mock, Mock, Run]:
+    """Simulate a run with a configuration without really launching the checks."""
+    configuration_path = str(configuration_path)
+    args = ["--rcfile", configuration_path, file_to_lint]
+    # If we used `pytest.raises(SystemExit)`, the `runner` variable
```
Comment (DanielNoord) [inline pylint/testutils/configuration_test.py:91]:
```suggestion
    # We do not capture the `SystemExit` as then the `runner` variable
```

File: pylint/testutils/configuration_test.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    # would not be accessible outside the `with` block.
+    with unittest.mock.patch("sys.exit") as mocked_exit:
+        # Do not actually run checks, that could be slow. Do not mock
```
Comment (DanielNoord) [inline pylint/testutils/configuration_test.py:94]:
```suggestion
        # Do not actually run checks, that could be slow. We don't mock
```

File: tests/config/test_functional_config_loading.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+Let's say you have a regression_list_crash.toml file to test. Then if there is an error in the conf,
+add ``regression_list_crash.out`` alongside your file with the expected output of pylint in it. Use
+``{relpath}`` and ``{abspath}`` for the path of the file. The exit code will have to be 2 (error)
+if this file exists.
+
+You must also define a ``regression_list_crash.result.json`` if you want to check the parsed configuration.
+This file will be loaded as a dict and will override the default value of the default pylint
+configuration. If you need to append or remove a value use the special key ``"functional_append"``
+and ``"functional_remove":``. Check the existing code for example.
```
Comment (DanielNoord) [inline tests/config/test_functional_config_loading.py:17]:
```suggestion
and ``"functional_remove":``. Check the existing code for examples.
```

File: tests/config/test_functional_config_loading.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    get_expected_configuration,
+    get_expected_output,
+    run_using_a_configuration_file,
+)
+
+HERE = Path(__file__).parent
+FUNCTIONAL_DIR = HERE / "functional"
+# We use string then recast to path so we can use -k in pytest otherwise
+# we get 'configuration_path0' as a test name. Relative to the functional
+# directory because otherwise the string is very lengthy
```
Comment (DanielNoord) [inline tests/config/test_functional_config_loading.py:39]:
```suggestion
# We use string then recast to path so we can use -k in pytest.
# Otherwise we get 'configuration_path0' as a test name. The path is relative to the functional
# directory because otherwise the string would be very lengthy.
```

File: tests/config/test_functional_config_loading.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+CONFIGURATION_PATHS = [
+    str(path.relative_to(FUNCTIONAL_DIR))
+    for ext in ("toml", "ini", "cfg")
```
Comment (DanielNoord) [inline tests/config/test_functional_config_loading.py:42]: Do we want to make this a global variable? `ACCEPTED_CONFIG_EXTENSIONS`?

File: tests/config/test_functional_config_loading.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    for path in FUNCTIONAL_DIR.rglob(f"*.{ext}")
+]
+FILE_TO_LINT = str(HERE / "file_to_lint.py")
+
+
+@pytest.fixture()
+def default_configuration(tmp_path: Path) -> Dict[str, Any]:
```
Comment (DanielNoord) [inline tests/config/test_functional_config_loading.py:49]: Comment for `Any`

File: tests/config/test_functional_config_loading.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+        str(empty_pylintrc), FILE_TO_LINT
+    )
+    mock_exit.assert_called_once_with(0)
+    return runner.linter.config.__dict__
+
+
+@pytest.mark.parametrize("configuration_path", CONFIGURATION_PATHS)
+def test_functional_config_loading(
+    configuration_path: str,
+    default_configuration: Dict[str, Any],
```
Comment (DanielNoord) [inline tests/config/test_functional_config_loading.py:62]: Comment for `Any`

File: pylint/testutils/configuration_test.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        # `Pylinter.check`: it calls `Pylinter.initialize` which is
+        # needed to properly set up messages inclusion/exclusion
+        # in `_msg_states`, used by `is_message_enabled`.
+        check = "pylint.lint.pylinter.check_parallel"
+        with unittest.mock.patch(check) as mocked_check_parallel:
+            runner = Run(args)
+    return mocked_exit, mocked_check_parallel, runner
```
Comment (DanielNoord) [inline pylint/testutils/configuration_test.py:101]: We never seem to use `mocked_check_parallel`. Do we need to return it?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Great work! I think the `functional_append` and `functional_remove` and not the easiest to understand for new contributors, but they probably also shouldn't need to mess with it too much.

Left some comments here and there but in general this seems quite polished!

File: pylint/testutils/configuration_test.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    def get_path_according_to_suffix() -> Path:
+        path = Path(pyproject_toml_path)
+        return path.parent / f"{path.stem}.{suffix}"
+
+    expected = default
+    expected_result_path = get_path_according_to_suffix()
+    if expected_result_path.exists():
+        with open(expected_result_path, encoding="utf8") as f:
+            expected = f.read()
+        logging.info("%s exists.", expected_result_path)
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:28]: No I don't think so. Here it's very helpful to see what's happening (especially to realize your file is not taken into account after a misspell of the file name). We can't print because we have a lot of writing into streams and the output of the program is checked during the test so printing messes with the result. I'll add a comment.

File: pylint/testutils/configuration_test.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+import json
+import logging
+import unittest
+from pathlib import Path
+from typing import Any, Dict, Tuple, Union
+from unittest.mock import Mock
+
+from pylint.lint import Run
+
+USER_SPECIFIC = str(Path(__file__).parent.parent.parent)
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:15]: ``/home/pierre/code/pylint/`` in ``/home/pierre/code/pylint/tests/config/functional/...``. It's not working on windows right now (see failing tests).

File: pylint/testutils/configuration_test.py (hunk 11 above)
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:101]: We could remove it, but this might permit to do additional check as it's a test util function that could be used by lib depending on pylint for their own tests, so I thought we might as well provide it.

File: pylint/testutils/configuration_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (DanielNoord) [reply pylint/testutils/configuration_test.py:28]: Ah, makes sense. Comment would be helpful indeed 😄

File: pylint/testutils/configuration_test.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (DanielNoord) [reply pylint/testutils/configuration_test.py:15]: I meant:
```suggestion
USER_SPECIFIC_PATH = str(Path(__file__).parent.parent.parent)
```
Sorry, should have made the suggestion immediately

File: pylint/testutils/configuration_test.py (hunk 3 above)
Comment (DanielNoord) [reply pylint/testutils/configuration_test.py:73]: Although duplicate I think a comment about why we're using `logging` here will help avoid future confusion.

File: pylint/testutils/configuration_test.py (hunk 11 above)
Comment (DanielNoord) [reply pylint/testutils/configuration_test.py:101]: Just wanted to check this was not left from an earlier version in which you did use the second return value 😄

Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:101]: In fact it's taken from [here](https://github.com/PyCQA/pylint/blob/main/tests/config/test_config.py#L15), I'll refactor and move everything to functional in a subsequent MR. Or maybe I should do it right now.

Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:102]: https://github.com/PyCQA/pylint/pull/5287/commits/7814c2fbb47a8a9f2824af77a7154201cd240964

File: tests/config/test_functional_config_loading.py (hunk 9 above)
Comment (Pierre-Sassoulas) [reply tests/config/test_functional_config_loading.py]: https://github.com/PyCQA/pylint/pull/5287/commits/050ffbf8f1806b20ee2bfc4e5994e558d094a373

File: tests/config/test_functional_config_loading.py (hunk 10 above)
Comment (Pierre-Sassoulas) [reply tests/config/test_functional_config_loading.py]: https://github.com/PyCQA/pylint/pull/5287/commits/050ffbf8f1806b20ee2bfc4e5994e558d094a373

File: pylint/testutils/configuration_test.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:47]: https://github.com/PyCQA/pylint/pull/5287/commits/050ffbf8f1806b20ee2bfc4e5994e558d094a373

File: tests/config/test_functional_config_loading.py (hunk 8 above)
Comment (Pierre-Sassoulas) [reply tests/config/test_functional_config_loading.py]: c6b66b0a82729f80cdd61dadd325ddb5e4081546

File: pylint/testutils/configuration_test.py (hunk 2 above)
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py]: 2eaad656d8c919dc9b007e1886e1d53d71fd69f4

File: pylint/testutils/configuration_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:37]: da4c8feb419a54544f076258bf4116bd8cd21fd3

File: pylint/testutils/configuration_test.py (hunk 3 above)
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:85]: da4c8feb419a54544f076258bf4116bd8cd21fd3

File: pylint/testutils/configuration_test.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def get_expected_or_default(pyproject_toml_path: str, suffix: str, default: Any) -> str:
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py]: 050ffbf8 and 0520fe43

File: pylint/config/option_manager_mixin.py [hunk 15]
```diff
@@ -271,6 +271,9 @@ def read_config_file(self, config_file=None, verbose=None):
 
         use_config_file = config_file and os.path.exists(config_file)
         if use_config_file:
+            mod_name = config_file
+            self.current_name = mod_name
+            self.set_current_module(mod_name)
```
Comment (Pierre-Sassoulas) [inline pylint/config/option_manager_mixin.py:276]: This is required to not crash on #4746 regression test.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Should be good to ~~go~~ review again 😄

File: tests/config/test_functional_config_loading.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+configuration. If you need to append or remove a value use the special key ``"functional_append"``
+and ``"functional_remove":``. Check the existing code for examples.
+"""
+
+# pylint: disable=redefined-outer-name
+import logging
+from pathlib import Path
+
+import pytest
+from _pytest.capture import CaptureFixture
```
Comment (DanielNoord) [inline tests/config/test_functional_config_loading.py:25]: There seems to be a `CaptureFixture` in the `pytest` namespace. Is there a difference? These seem to be private classes.

**Review by DanielNoord — COMMENTED** [review]
Minor comments here and there.

Should we also test a `pylintrc` without any extension in this test framework?

File: tests/config/functional/toml/issue_4746/loaded_plugin_does_not_exists.toml [hunk 17]
```diff
@@ -0,0 +1,8 @@
+# The pylint_websockets plugin does not exist and therefore this toml is invalid
+[tool.poe.tasks]
+docs = {cmd = "sphinx-build docs build", help = "Build documentation"}
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply tests/config/functional/toml/issue_4746/loaded_plugin_does_not_exists.toml:3]: It's only the example in the original issue, I will remove the unrequired options.

File: tests/config/test_functional_config_loading.py (hunk 16 above)
Comment (Pierre-Sassoulas) [reply tests/config/test_functional_config_loading.py:25]: It's a pycharm suggestion, it probably missed the right import.

**Comment by Pierre-Sassoulas** [comment]
> Should we also test a `pylintrc` without any extension in this test framework?

Pylintrc is an ini file, I think the file name itself does not matter a lot. (The 'find the  conf file' logic is handled elsewhere)

File: pylint/testutils/configuration_test.py [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+
+EXPECTED_CONF_APPEND_KEY = "functional_append"
+EXPECTED_CONF_REMOVE_KEY = "functional_remove"
+
+
+def get_expected_configuration(
+    configuration_path: str, default_configuration: PylintConfiguration
+) -> Dict[str, Any]:
+    """Get the expected parsed configuration of a configuration functional test"""
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylint/testutils/configuration_test.py:52]: 8813782c

File: tests/config/functional/toml/issue_4746/loaded_plugin_does_not_exists.toml (hunk 17 above)
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply tests/config/functional/toml/issue_4746/loaded_plugin_does_not_exists.toml]: 8154ae61

File: tests/config/test_functional_config_loading.py (hunk 16 above)
Comment (Pierre-Sassoulas) [reply tests/config/test_functional_config_loading.py]: c166b28a

File: tests/lint/test_pylinter.py [hunk 19]
```diff
@@ -2,9 +2,9 @@
 from typing import Any
 from unittest.mock import patch
 
-from _pytest.capture import CaptureFixture
 from astroid import AstroidBuildingError
 from py._path.local import LocalPath  # type: ignore
+from pytest import CaptureFixture
```
Comment (DanielNoord) [inline tests/lint/test_pylinter.py:7]: 👍

File: pylint/config/option_manager_mixin.py (hunk 15 above)
Comment (Pierre-Sassoulas) [reply pylint/config/option_manager_mixin.py:276]: Is this what you're suggesting to do ?
````diff
diff --git a/pylint/config/option_manager_mixin.py b/pylint/config/option_manager_mixin.py
index 924998e7..3595d466 100644
--- a/pylint/config/option_manager_mixin.py
+++ b/pylint/config/option_manager_mixin.py
@@ -271,9 +271,8 @@ class OptionsManagerMixIn:
 
         use_config_file = config_file and os.path.exists(config_file)
         if use_config_file:
-            mod_name = config_file
-            self.current_name = mod_name
-            self.set_current_module(mod_name)
+            self.current_name = config_file
+            self.set_current_module(self.current_name)
             parser = self.cfgfile_parser
             if config_file.endswith(".toml"):
                 self._parse_toml(config_file, parser)
```

File: pylint/config/option_manager_mixin.py [hunk 20]
```diff
@@ -271,6 +271,8 @@ def read_config_file(self, config_file=None, verbose=None):
 
         use_config_file = config_file and os.path.exists(config_file)
         if use_config_file:
+            self.current_name = config_file
+            self.set_current_module(self.current_name)
```
Comment (DanielNoord) [inline pylint/config/option_manager_mixin.py:275]:
```suggestion
            self.set_current_module(config_file)
```

This should work as well as `set_current_module` sets the `current_name`.

**Review by DanielNoord — APPROVED** [review]

File: pylint/config/option_manager_mixin.py (hunk 15 above)
Comment (DanielNoord) [reply pylint/config/option_manager_mixin.py]: See my latest comment!

File: pylint/config/option_manager_mixin.py (hunk 20 above)
Comment (Pierre-Sassoulas) [reply pylint/config/option_manager_mixin.py]: Ha, right, thank you for clarifying !

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas Did you test this with `tox`? The test break for me locally when using `tox`. `pytest` works fine.

**Comment by Pierre-Sassoulas** [comment]
No I'm using only pytest but I thought github action used tox. Let me check.

**Comment by Pierre-Sassoulas** [comment]
See https://github.com/PyCQA/pylint/pull/5301

File: pylint/config/option_manager_mixin.py [hunk 21]
```diff
@@ -271,6 +271,7 @@ def read_config_file(self, config_file=None, verbose=None):
 
         use_config_file = config_file and os.path.exists(config_file)
         if use_config_file:
+            self.set_current_module(config_file)
```
Comment (DanielNoord) [inline pylint/config/option_manager_mixin.py:274]: @Pierre-Sassoulas Why is this necessary? Does this add anything other than allowing the test framework to function?

_Ran into this while working on `argparse`._

Comment (Pierre-Sassoulas) [reply pylint/config/option_manager_mixin.py:274]: I think it's setting the file that the ``add_message`` will display if we add a message later. I.E. ``[relpath}:1:0: E0013:`` later.
